runtime-sdk/src/dispatcher.rs

```diff
@@ -467,8 +440,130 @@ impl<R: Runtime + Send + Sync> transaction::dispatcher::Dispatcher for Dispatche
+                    results.push(Self::execute_tx(ctx, tx_size, tx, index)?);
+                }
+
+                Ok(results)
+            },
+        )
+    }
+
+    fn schedule_and_execute_batch(
+        &self,
+        rt_ctx: transaction::Context<'_>,
+        batch: &mut TxnBatch,
+    ) -> Result<ExecuteBatchResult, RuntimeError> {
+        let cfg = R::SCHEDULE_CONTROL.unwrap(); // Must succeed otherwise we wouldn't be here.
+        let mut tx_reject_hashes = Vec::new();
+
+        let mut result = self.execute_batch_common(
+            rt_ctx,
+            |ctx| -> Result<Vec<ExecuteTxResult>, RuntimeError> {
+                // Schedule and execute the batch.
+                //
+                // The idea is to keep scheduling transactions as long as we have some space
+                // available in the block as determined by gas use.
+                let mut new_batch = Vec::new();
+                let mut results = Vec::with_capacity(batch.len());
+                'batch: loop {
+                    for raw_tx in batch.drain(..) {
+                        // If we don't have enough gas for processing even the cheapest transaction
+                        // we are done.
+                        let remaining_gas = R::Core::remaining_batch_gas(ctx);
+                        if remaining_gas < cfg.min_remaining_gas {
+                            break 'batch;
+                        }
+
+                        // Decode transaction.
+                        let tx = match Self::decode_tx(ctx, &raw_tx) {
+                            Ok(tx) => tx,
+                            Err(_) => {
+                                // Transaction is malformed, make sure it gets removed from the
+                                // queue and don't include it in a block.
+                                tx_reject_hashes.push(Hash::digest_bytes(&raw_tx));
+                                continue;
+                            }
+                        };
+                        let tx_size = raw_tx.len().try_into().unwrap();
+
+                        // If we don't have enough gas remaining to process this transaction, just
+                        // skip it.
+                        if tx.auth_info.fee.gas > remaining_gas {
+                            continue;
+                        }
+                        // Same if we don't have enough consensus message slots.
+                        if tx.auth_info.fee.consensus_messages > ctx.remaining_messages() {
+                            continue;
+                        }
+
+                        // Determine the current transaction index.
+                        let index = new_batch.len();
+
+                        new_batch.push(raw_tx);
+                        results.push(Self::execute_tx(ctx, tx_size, tx, index)?);
+                    }
+
+                    // TODO: If there's more room in the block and we got the maximum number of
```

Comment:
Also some timeout probably? So it doesn't indefinitely wait for new transactions?

Reply:
Yeah so the idea is to do it until either the batch is full (remaining batch gas is below the set minimum) or we've exhausted the queue (returned length is less than the requested length).

Reply:
This now has an implementation, please take a look.

Reply:
Looks good